src/apps/bristleback_apps/serial_payload_example/user_code/user_code.cpp

```diff
@@ -36,44 +62,94 @@ void setup(void) {
-    if (!led_state) {
-      IOWrite(&LED_GREEN, 0);
-      led_state = true;
-    } else {
-      IOWrite(&LED_GREEN, 1);
-      led_state = false;
+    // Shortcut the raw bytes cluster completion so the parsed line will be on a new console line
+    if (readingBytesTimer > -1) {
+      printf("\n");
+      readingBytesTimer = -1;
     }
-
-    uint16_t read_len =
-        PLUART::readLine(payload_buffer, sizeof(payload_buffer));
-    printf("line: %s\n", payload_buffer);
+    uint16_t read_len = PLUART::readLine(payload_buffer, sizeof(payload_buffer));
 
     // Get the RTC if available
     RTCTimeAndDate_t time_and_date = {};
     rtcGet(&time_and_date);
     char rtcTimeBuffer[32];
     rtcPrint(rtcTimeBuffer, &time_and_date);
 
-    // Print the payload data to a file, to the bm_printf console, and to the printf console.
-    bm_fprintf(0, "payload_data.log", USE_TIMESTAMP, "tick: %llu, rtc: %s, line: %.*s\n",
-               uptimeGetMs(), rtcTimeBuffer, read_len, payload_buffer);
-    bm_printf(0, "[payload] | tick: %llu, rtc: %s, line: %.*s", uptimeGetMs(),
-              rtcTimeBuffer, read_len, payload_buffer);
-    printf("[payload] | tick: %llu, rtc: %s, line: %.*s\n", uptimeGetMs(),
-           rtcTimeBuffer, read_len, payload_buffer);
+    // Based on configuration, print the payload data to a file, to the bm_printf console, and to the printf console.
+    if (bm_log_enable) {
+      bm_fprintf(0, "payload_data.log", USE_TIMESTAMP, "tick: %llu, rtc: %s, line: %.*s\n",
+                 uptimeGetMs(), rtcTimeBuffer, read_len, payload_buffer);
+      bm_printf(0, "[payload] | tick: %llu, rtc: %s, line: %.*s", uptimeGetMs(), rtcTimeBuffer, read_len, payload_buffer);
+    }
+      printf("[payload-line] | tick: %llu, rtc: %s, line: %.*s\n", uptimeGetMs(), rtcTimeBuffer, read_len, payload_buffer);
+
+    ledLinePulse = uptimeGetMs(); // trigger a pulse on LED
+  }
+
+  static bool led2State = false;
+  /// Flash the LED blue if we received Rx data
+  if (!led2State && ledLinePulse > -1) {
+    IOWrite(&LED_BLUE, 1);
+    led2State = true;
+  }
+  // If LED2 has been on for LED_ON_TIME_MS, turn it off.
+  else if (led2State && ((u_int32_t)uptimeGetMs() - ledLinePulse >= LED_ON_TIME_MS)) {
+    IOWrite(&LED_BLUE, 0);
+    ledLinePulse = -1;
+    led2State = false;
+  }
+
+  /// Blink the LED green periodically for sign of life.
+  static u_int32_t ledPulseTimer = uptimeGetMs();
+  static u_int32_t ledOnTimer = 0;
+  static bool led1State = false;
+  // Turn LED1 on green every LED_PERIOD_MS milliseconds.
+  if (!led1State && ((u_int32_t)uptimeGetMs() - ledPulseTimer >= LED_PERIOD_MS)) {
+    IOWrite(&LED_GREEN, 1);
+    ledOnTimer = uptimeGetMs();
+    ledPulseTimer += LED_PERIOD_MS;
+    led1State = true;
+  }
+  // If LED1 has been on for LED_ON_TIME_MS milliseconds, turn it off.
+  else if (led1State && ((u_int32_t)uptimeGetMs() - ledOnTimer >= LED_ON_TIME_MS)) {
+    IOWrite(&LED_GREEN, 0);
+    led1State = false;
```

Comment:
Could these and maybe some other chunks of code be broken down into functions (even inline functions), just to add more clarity, and improve readability?
If anything the auxiliary functions like blinking LEDs might help reduce some of the eye clutter.
i.e.
```
payload_print_to_file(...);

receive_rx_led_blink_blue();

heartbeat_led_green()
```